tests/src/Functional/AccessTest.php

```diff
@@ -0,0 +1,251 @@
+            [
+                'preview public_announcement in collabora',
+                'preview public_wiki in collabora',
+                'edit any public_wiki in collabora',
+            ],
+        );
+        $diary_keeper = $this->createUser([
+            'preview diary in collabora',
+            'edit own diary in collabora'
+        ]);
+
+        $media_entities = [
+            'document' => $this->createMediaEntity('document'),
+            'wiki' => $this->createMediaEntity('public_wiki'),
+            'announcement' => $this->createMediaEntity('public_announcement'),
+            'own diary' => $this->createMediaEntity('diary', ['uid' => $diary_keeper->id()]),
+            'other diary' => $this->createMediaEntity('diary'),
+        ];
+
+        $paths = [];
+        foreach ($media_entities as $media_key => $media) {
+            $paths["/cool/view/<$media_key>"] = '/cool/view/' . $media->id();
+            $paths["/cool/edit/<$media_key>"] = '/cool/edit/' . $media->id();
+        }
+        $users = [
+            'anonymous' => new AnonymousUserSession(),
+            'authenticated' => $this->createUser(),
+            'reader' => $this->createUser([
+                'preview document in collabora',
+            ]),
+            'editor' => $this->createUser([
+                'preview document in collabora',
+                'edit any document in collabora',
+            ]),
+            // Usually, a user with write access would also be given read access.
+            'writer' => $this->createUser([
+                'edit any document in collabora',
+            ]),
+            'diary keeper' => $diary_keeper,
+        ];
+
+        // Both routes are only accessible for admin.
+        $this->assertSameReport(
+            [
+                '/cool/view/<document>' => ['reader', 'editor'],
+                '/cool/edit/<document>' => ['editor', 'writer'],
+                '/cool/view/<wiki>' => ['anonymous'],
+                '/cool/edit/<wiki>' => ['anonymous'],
+                '/cool/view/<announcement>' => ['anonymous'],
+                '/cool/edit/<announcement>' => [],
+                '/cool/view/<own diary>' => ['diary keeper'],
+                '/cool/edit/<own diary>' => ['diary keeper'],
+                '/cool/view/<other diary>' => ['diary keeper'],
+                '/cool/edit/<other diary>' => [],
+            ],
+            $this->buildAccessReport($paths, $users),
+            'Users with access to given paths',
+        );
+    }
+
+    /**
+     * Builds a report about which users can access a given content.
+     *
+     * @param list<string>|array<string, string> $paths
+     *   A list of paths, optionally keyed.
+     * @param array<string, \Drupal\Core\Session\AccountInterface> $users
+     *   Accounts to test access with.
+     *
+     * @return array
+     *   Report to compare against in assertions.
+     */
+    protected function buildAccessReport(array $paths, array $users): array {
```

Comment:
this is too cryptic for me (and prob most people), in the test I need to see clearly that role x and y have or not permission for routes. I can guess what this `'/cool/view/<document>' => ['reader', 'editor'],` means but dont really know until I compare what `buildAccessReport` is actually composing

Reply:
The benefit of the "compare big array" is that you see in one big overview what you need to fix.
if we do separate assertions, we only see the first failure which would say "user x has no access to path y".
knowing which users have access and which don't is quite useful for troubleshooting.

Could we resolve this confusion with comments or improved assertion message?
Or is the problem in the "magicalness" of the `buildAccessReport()`?

Reply:
magicalness yes, need to be more readable without debuging